feat: Loading builder for Route

PistonShot1 opened the pull request:
<!-- Exclude from commit message -->
<!--
The title of your PR on the line above should start with a [Conventional Commit] prefix
(`fix:`, `feat:`, `docs:`, `test:`, `chore:`, `refactor:`, `perf:`, `build:`, `ci:`,
`style:`, `revert:`). This title will later become an entry in the [CHANGELOG], so please
make sure that it summarizes the PR adequately.

Don't remove the "exclude from commit message" comments below. They are used to prevent
the PR description template from being included in the git log.
Only change the "Replace this text" parts.
-->

# Description
<!--
Provide a description of what this PR is doing.
If you're modifying existing behavior, describe the existing behavior, how this PR is changing it,
and what motivated the change. If this is a breaking change, specify explicitly which APIs were
changed.
-->
<!-- End of exclude from commit message -->
This change is mainly to make it easier for specifying loading screen along when routing between component using the ``RouterComponent``. Currently the change is only made on ``Route``, that now have optional ``loadingBuilder`` to Route constructor which would expect a ``Component``, which would be added as the loading page/screen while  ``onLoad`` of the ``builder`` (the main routing component) to be completed.

<!-- Exclude from commit message -->
## Checklist
<!--
Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes with `[x]`. If some checkbox is not applicable, mark it as `[-]`.
-->

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples` or `docs`.


## Breaking Change?
<!--
Would your PR require Flame users to update their apps following your change?

If yes, then the title of the PR should include "!" (for example, `feat!:`, `fix!:`). See
[Conventional Commit] for details. Also, for a breaking PR uncomment and fill in the "Migration
instructions" section below.
-->

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

<!--
### Migration instructions

If the PR is breaking, uncomment this header and add instructions for how to migrate from the
currently released version in-between the two following tags:
-->
<!-- End of exclude from commit message -->
<!-- Exclude from commit message -->


<!-- Links -->
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md
<!-- End of exclude from commit message -->

spydon commented on packages/flame/lib/src/components/route.dart:
```diff
@@ -118,17 +125,31 @@ class Route extends PositionComponent
 
+  /// The loadingpage that was built and is now owned by this route. This
+  /// loadingpage will also be added as a child component.
+  Component? _loadingpage;
```
```suggestion
  Component? _loadingPage;
```

spydon replied:
The spell checker step in the pipeline will also be happy once this is updated.

spydon reviewed (approved):
Lgtm, very clean implementation.
Could you add a simple test that just tests that both the loadingPage and then the page is shown?
It would be in here:
https://github.com/flame-engine/flame/blob/main/packages/flame/test/components/route_test.dart

PistonShot1 commented:
wait ya , dont merge yet, adding test for this

PistonShot1 commented:
Don't merge yet.

hi @spydon , sorry I was away for awhile. Just pushed new one with test. Anyways, I just have problem it making a full test to check if the loading component was mounted before it was removed. I was only able to check for loading component was removed and the builder component ( page component) was mounted right after it was removed.

If you look at the file route_test.dart , I added test and also did add few comments , to explain the situation for the test coverage. Sorry I am not that good at coming with test, but I did manually check it by linking my version of flame repo to one of my old flame game, I mean my flame in pubspec refer to my git repo instead of pub.dev and test this feature I added, according to my observation it did work to what I expected.

PistonShot1 commented:
I mean does checking the loading component was removed technically means it also verifies that it was mounted before right . So if that is the case then the test I added already suffice ?

spydon commented on packages/flame/test/components/route_test.dart:
```diff
@@ -474,6 +475,51 @@ void main() {
+      // following check either, that is to check state when loading was 
+      // mounted before it was removed.
+
+      // expect(
+      //   loadingComponent.isMounted,
+      //   isTrue,
+      // );
+      // expect(
+      //   pageComponent.isMounted,
+      //   isFalse,
+      // );
```
I don't think I follow why this doesn't work, what are you getting here currently?

PistonShot1 replied:
I'm sorry , False is being expected for 503, as for 507 it does pass

spydon replied:
Why is false expected for 503? Since the route is added it should be showing the `loadingComponent` right?

PistonShot1 replied:
I just found, I made an error in the checking at line 514 is not accurate in this commit, it should be expecting True, but I checked there for False and it passed

That means my loading component was never mounted even before and after the main component was mounted? 
Maybe, that mean my implementation may have some error ?

spydon replied:
I'm guessing that due to how `Future.delayed` works in tests the `loadingComponent` was added and removed in the same tick. You could pass in a future to the `_HeavyComponent` that you manually complete after you have checked that the `loadingComponent` is mounted.

spydon replied:
Did you have any more time to look at this @PistonShot1?

PistonShot1 replied:
Sorry for the delays got caught up with another project, but I am trying to implement the approach you mentioned , to pass a future that I can manually complete, I am not exactly sure how to implement this . I am assuming with `Completer` or sort ?

spydon replied:
Yeah, `Completer` should work

PistonShot1 replied:
the test for checking if the `loadingComponent` was mounted and remove, both doesn't pass (at line 499 and at line 509), as I have commented, it is some how implying that the loadingPage was never mounted. I am not sure why, but the changes I added at `packages/flame/lib/src/components/route.dart` (line 146-151), supposed have mounted and removed after awaiting `_page!.loaded;`

spydon commented:
@PistonShot1 have you made any progress on this? :)

PistonShot1 commented:
@spydon . I have fixed it, sorry took so long for such a small feature addition, I gave up after the test keep on fail, but as you suspected it was the way I wrote the test, now it works fine and should be working and passed the expected tests

spydon commented on packages/flame/lib/src/components/route.dart:
```diff
@@ -53,6 +55,11 @@ class Route extends PositionComponent
   /// in which case the user must override the [build] method.
   final Component Function()? _builder;
 
+  /// The function that will be invoked that will build the loading page
+  /// component when this route first becomes active. Can be left `null`,
+  /// in which case it is not required.
```
```suggestion
  /// The function that will build the loading page component, which is shown
  /// when this route first becomes active, but hasn't fully loaded yet.
```

spydon commented on packages/flame/lib/src/components/route.dart:
```diff
@@ -118,17 +125,32 @@ class Route extends PositionComponent
   /// also be added as a child component.
   Component? _page;
 
+  /// The loadingPage that was built and is now owned by this route. This
+  /// loadingPage will also be added as a child component.
```
```suggestion
  /// The loadingPage that was built and is now owned by this route. The
  /// [_loadingPage] will also be added as a child component.
```

spydon commented on packages/flame/lib/src/components/route.dart:
```diff
@@ -118,17 +125,32 @@ class Route extends PositionComponent
+  Future<void> _addLoadingPage() async {
+    _loadingPage ??= _loadingBuilder!()..addToParent(this);
+    await _loadingPage!.loaded;
+    await _page!.addToParent(this);
```
```suggestion
    await add(_page!);
```

spydon reviewed (approved):
Lgtm!
